File: backend/app/api/stats.py
"""
平台统计 API
"""
from fastapi import APIRouter, Depends
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models import Blueprint, User, Favorite, Like, Report
from app.schemas import StatsResponse

router = APIRouter(prefix="/api", tags=["stats"])


@router.get("/stats", response_model=StatsResponse)
async def get_stats(db: AsyncSession = Depends(get_db)):
    """返回平台统计数据（公开）。"""
    total_blueprints = (await db.execute(
        select(func.count()).select_from(Blueprint).where(Blueprint.is_published == True)
    )).scalar() or 0

    total_users = (await db.execute(
        select(func.count()).select_from(User)
    )).scalar() or 0

    total_favorites = (await db.execute(
        select(func.count()).select_from(Favorite)
    )).scalar() or 0

    total_pieces = (await db.execute(
        select(func.coalesce(func.sum(Blueprint.piece_count), 0))
        .where(Blueprint.is_published == True)
    )).scalar() or 0

    total_views = (await db.execute(
        select(func.coalesce(func.sum(Blueprint.view_count), 0))
    )).scalar() or 0

    total_likes = (await db.execute(
        select(func.count()).select_from(Like)
    )).scalar() or 0

    pending_count = (await db.execute(
        select(func.count()).select_from(Blueprint).where(Blueprint.is_published == False)
    )).scalar() or 0

    report_count = (await db.execute(
        select(func.count(func.distinct(Report.blueprint_id)))
    )).scalar() or 0

    return StatsResponse(
        total_blueprints=total_blueprints,
        total_users=total_users,
        total_favorites=total_favorites,
        total_pieces=total_pieces,
        total_views=total_views,
        total_likes=total_likes,
        pending_count=pending_count,
        report_count=report_count,
    )

File: backend-go/internal/handler/stats.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	"brickplans/internal/config"
	"brickplans/internal/db"
)

type StatsHandler struct {
	cfg *config.Config
	gdb *gorm.DB
}

func NewStatsHandler(cfg *config.Config, gdb *gorm.DB) *StatsHandler {
	return &StatsHandler{cfg: cfg, gdb: gdb}
}

func (h *StatsHandler) RegisterRoutes(rg *gin.RouterGroup) {
	rg.GET("/stats", h.get)
}

func (h *StatsHandler) get(c *gin.Context) {
	var totalBlueprints, totalUsers, totalFavorites, totalLikes, pending, reportCount int64
	var totalPieces, totalViews int64

	h.gdb.Model(&db.Blueprint{}).Where("is_published = ?", true).Count(&totalBlueprints)
	h.gdb.Model(&db.User{}).Count(&totalUsers)
	h.gdb.Model(&db.Favorite{}).Count(&totalFavorites)
	h.gdb.Model(&db.Like{}).Count(&totalLikes)
	h.gdb.Model(&db.Blueprint{}).Where("is_published = ?", false).Count(&pending)
	h.gdb.Model(&db.Report{}).Distinct("blueprint_id").Count(&reportCount)

	h.gdb.Model(&db.Blueprint{}).Where("is_published = ?", true).
		Select("COALESCE(SUM(piece_count), 0)").Scan(&totalPieces)
	h.gdb.Model(&db.Blueprint{}).
		Select("COALESCE(SUM(view_count), 0)").Scan(&totalViews)

	c.JSON(http.StatusOK, gin.H{
		"total_blueprints": totalBlueprints,
		"total_users":      totalUsers,
		"total_favorites":  totalFavorites,
		"total_pieces":     totalPieces,
		"total_views":      totalViews,
		"total_likes":      totalLikes,
		"pending_count":    pending,
		"report_count":     reportCount,
	})
}
